In [3]:
import pandas as pd
url = "https://data.cityofnewyork.us/api/views/vfnx-vebw/rows.csv?accessType=DOWNLOAD"
df = pd.read_csv(url)
df

X          Y Unique Squirrel ID Hectare Shift      Date  \
0    -73.956134  40.794082     37F-PM-1014-03     37F    PM  10142018   
1    -73.968857  40.783783     21B-AM-1019-04     21B    AM  10192018   
2    -73.974281  40.775534     11B-PM-1014-08     11B    PM  10142018   
3    -73.959641  40.790313     32E-PM-1017-14     32E    PM  10172018   
4    -73.970268  40.776213     13E-AM-1017-05     13E    AM  10172018   
...         ...        ...                ...     ...   ...       ...   
3018 -73.963943  40.790868     30B-AM-1007-04     30B    AM  10072018   
3019 -73.970402  40.782560     19A-PM-1013-05     19A    PM  10132018   
3020 -73.966587  40.783678     22D-PM-1012-07     22D    PM  10122018   
3021 -73.963994  40.789915     29B-PM-1010-02     29B    PM  10102018   
3022 -73.975479  40.769640      5E-PM-1012-01     05E    PM  10122018   

      Hectare Squirrel Number    Age Primary Fur Color  \
0                           3    NaN               NaN   
1                           4    NaN               NaN   
2                           8    NaN              Gray   
3                          14  Adult              Gray   
4                           5  Adult              Gray   
...                       ...    ...               ...   
3018                        4  Adult              Gray   
3019                        5  Adult              Gray   
3020                        7  Adult              Gray   
3021                        2    NaN              Gray   
3022                        1  Adult          Cinnamon   

         Highlight Fur Color  ...   Kuks  Quaas  Moans Tail flags  \
0                        NaN  ...  False  False  False      False   
1                        NaN  ...  False  False  False      False   
2                        NaN  ...  False  False  False      False   
3                        NaN  ...  False  False  False      False   
4                   Cinnamon  ...  False  False  False      False   
...                      ...  ...    ...    ...    ...        ...   
3018                     NaN  ...  False  False  False      False   
3019                   White  ...  False  False  False      False   
3020  Black, Cinnamon, White  ...  False  False  False      False   
3021         Cinnamon, White  ...  False  False  False      False   
3022             Gray, White  ...  False  False  False      False   

     Tail twitches  Approaches  Indifferent  Runs from  Other Interactions  \
0            False       False        False      False                 NaN   
1            False       False        False      False                 NaN   
2            False       False        False      False                 NaN   
3            False       False        False       True                 NaN   
4            False       False        False      False                 NaN   
...            ...         ...          ...        ...                 ...   
3018         False       False        False       True                 NaN   
3019         False       False         True      False                 NaN   
3020         False       False         True      False                 NaN   
3021         False       False         True      False                 NaN   
3022         False        True        False      False                 NaN   

                                        Lat/Long  
0     POINT (-73.9561344937861 40.7940823884086)  
1     POINT (-73.9688574691102 40.7837825208444)  
2     POINT (-73.97428114848522 40.775533619083)  
3     POINT (-73.9596413903948 40.7903128889029)  
4     POINT (-73.9702676472613 40.7762126854894)  
...                                          ...  
3018  POINT (-73.9639431360458 40.7908677445466)  
3019  POINT (-73.9704015859639 40.7825600069973)  
3020  POINT (-73.9665871993517 40.7836775064883)  
3021  POINT (-73.9639941227864 40.7899152327912)  
3022  POINT (-73.9754794191553 40.7696404489025)  

[3023 rows x 31 columns]

In [5]:
# Get the number of rows and columns
num_rows, num_columns = df.shape

print(f'Number of rows: {num_rows}')
print(f'Number of columns: {num_columns}')

Number of rows: 3023
Number of columns: 31


Question 2.2:

Variables the features that are being recorded within a DataFrame, or the columns. For this particular dataset, they represent attributes such as fur colour, age, tail twitches, etc.

Observations are the individual pieces of data within a DataFrame, or the rows. For this particular dataset, they represent the specific data pieces for each individual squirrel under each variable

In [7]:
print(df.describe())

                 X            Y          Date  Hectare Squirrel Number
count  3023.000000  3023.000000  3.023000e+03              3023.000000
mean    -73.967184    40.780853  1.011949e+07                 4.123718
std       0.007726     0.010285  4.246671e+04                 3.096492
min     -73.981159    40.764911  1.006202e+07                 1.000000
25%     -73.973102    40.771676  1.008202e+07                 2.000000
50%     -73.968594    40.778166  1.012202e+07                 3.000000
75%     -73.960189    40.791219  1.014202e+07                 6.000000
max     -73.949722    40.800119  1.020202e+07                23.000000


In [8]:
for column in df.select_dtypes(include=['object']).columns:
    print(f'\nSummary for {column}:')
    print(df[column].value_counts())


Summary for Unique Squirrel ID:
Unique Squirrel ID
7D-PM-1010-01     2
40B-AM-1019-06    2
37E-PM-1006-03    2
4C-PM-1010-05     2
1F-AM-1010-04     2
                 ..
13F-AM-1007-01    1
12D-AM-1013-02    1
2F-AM-1006-03     1
9H-PM-1018-04     1
5E-PM-1012-01     1
Name: count, Length: 3018, dtype: int64

Summary for Hectare:
Hectare
14D    32
32E    30
14E    28
01B    27
07H    26
       ..
11C     1
23D     1
26B     1
24B     1
30F     1
Name: count, Length: 339, dtype: int64

Summary for Shift:
Shift
PM    1676
AM    1347
Name: count, dtype: int64

Summary for Age:
Age
Adult       2568
Juvenile     330
?              4
Name: count, dtype: int64

Summary for Primary Fur Color:
Primary Fur Color
Gray        2473
Cinnamon     392
Black        103
Name: count, dtype: int64

Summary for Highlight Fur Color:
Highlight Fur Color
Cinnamon                  767
White                     585
Cinnamon, White           268
Gray                      170
Gray, White                59
Black

In [9]:
df.shape

(3023, 31)

In [10]:
df.describe()

X            Y          Date  Hectare Squirrel Number
count  3023.000000  3023.000000  3.023000e+03              3023.000000
mean    -73.967184    40.780853  1.011949e+07                 4.123718
std       0.007726     0.010285  4.246671e+04                 3.096492
min     -73.981159    40.764911  1.006202e+07                 1.000000
25%     -73.973102    40.771676  1.008202e+07                 2.000000
50%     -73.968594    40.778166  1.012202e+07                 3.000000
75%     -73.960189    40.791219  1.014202e+07                 6.000000
max     -73.949722    40.800119  1.020202e+07                23.000000

Question 4:

a) The reason why there is a discrepency between the amount columns for df.shape and df.describe() is because the describe() method only analyzes columns with numerical values for the purpose of generating statistical information, resulting in the non-numerical columns being ejected from the table.

b) The values being reported in the count row of each column represents how many non-null entries reside within that column in the original DataFrame. Since for the squirrel DataFrame there were no null entries in the numerical columns, the count is equivalent to the amount of rows.

Question 5:

The difference between attributes and methods are the way in which they interact with objects.

Attributes do not modify any data within an object but rather represents the various properties of an object. The shape attribute of the pandas DataFrame object, for example, examines and displays the dimensions of a given DataFrame.

Methods, on the other hand, modifies and calculates the data stored within an object for the purpose of completing some other task. The describe() method of the pandas DataFrame, for example, extracts numerical data from a given DataFrame for the purpose of calculating and displaying the particular statistics.

Chatbot Session 1:

Summary
Dataset Insights:

You provided a link to a dataset of detailed observations of squirrels found in Central Park and asked about determining the number of rows and columns in the dataset.
We discussed using the df.shape attribute to get the number of rows and columns and df.describe() for summary statistics of numeric columns.
Data Discrepancy:

You noted a discrepancy between the number of rows and columns reported by df.shape and df.describe().
We explained that df.describe() provides statistics for numeric columns only and might not include all rows if there are missing values.
Attributes vs. Methods:

We discussed the difference between attributes and methods in Python:
Attributes: Variables that store data or state related to an object. Example: df.shape.
Methods: Functions that perform operations or computations. Example: df.describe().
Python Objects:

We clarified that df is an object in Python, specifically an instance of the pandas.DataFrame class.
As an object, df has attributes (like shape) and methods (like describe()) that are defined by the DataFrame class.
Feel free to reach out if you have any more questions or need further assistance!

https://chatgpt.com/share/8c468e40-2624-4487-b78d-83e63ae236a5

Question 6:

Count: Represents the amount of non-null (filled) rows for each column from a given DataFrame.

Mean: The average value of the rows within a column of a given DataFrame

std: The standard deviation of a given DataFrame. Measures the dispersion of values relative to the mean.

min: The minimum value of a column.

25%: The 25th percentile value of a column. Represents the value that a quarter of the data falls under.

50%: The median value AKA the 50th percentile. Separates the data of a column into two halves.

75%: The 75th percentile value of a column. Represents the value that 3/4th's of the data falls under.

max: the highest value of a column.

In [88]:
df_original = df.copy()

In [89]:
df_original

X          Y Unique Squirrel ID Hectare Shift      Date  \
0    -73.956134  40.794082     37F-PM-1014-03     37F    PM  10142018   
1    -73.968857  40.783783     21B-AM-1019-04     21B    AM  10192018   
2    -73.974281  40.775534     11B-PM-1014-08     11B    PM  10142018   
3    -73.959641  40.790313     32E-PM-1017-14     32E    PM  10172018   
4    -73.970268  40.776213     13E-AM-1017-05     13E    AM  10172018   
...         ...        ...                ...     ...   ...       ...   
3018 -73.963943  40.790868     30B-AM-1007-04     30B    AM  10072018   
3019 -73.970402  40.782560     19A-PM-1013-05     19A    PM  10132018   
3020 -73.966587  40.783678     22D-PM-1012-07     22D    PM  10122018   
3021 -73.963994  40.789915     29B-PM-1010-02     29B    PM  10102018   
3022 -73.975479  40.769640      5E-PM-1012-01     05E    PM  10122018   

      Hectare Squirrel Number    Age Primary Fur Color  \
0                           3    NaN               NaN   
1                           4    NaN               NaN   
2                           8    NaN              Gray   
3                          14  Adult              Gray   
4                           5  Adult              Gray   
...                       ...    ...               ...   
3018                        4  Adult              Gray   
3019                        5  Adult              Gray   
3020                        7  Adult              Gray   
3021                        2    NaN              Gray   
3022                        1  Adult          Cinnamon   

         Highlight Fur Color  ...   Kuks  Quaas  Moans Tail flags  \
0                        NaN  ...  False  False  False      False   
1                        NaN  ...  False  False  False      False   
2                        NaN  ...  False  False  False      False   
3                        NaN  ...  False  False  False      False   
4                   Cinnamon  ...  False  False  False      False   
...                      ...  ...    ...    ...    ...        ...   
3018                     NaN  ...  False  False  False      False   
3019                   White  ...  False  False  False      False   
3020  Black, Cinnamon, White  ...  False  False  False      False   
3021         Cinnamon, White  ...  False  False  False      False   
3022             Gray, White  ...  False  False  False      False   

     Tail twitches  Approaches  Indifferent  Runs from  Other Interactions  \
0            False       False        False      False                 NaN   
1            False       False        False      False                 NaN   
2            False       False        False      False                 NaN   
3            False       False        False       True                 NaN   
4            False       False        False      False                 NaN   
...            ...         ...          ...        ...                 ...   
3018         False       False        False       True                 NaN   
3019         False       False         True      False                 NaN   
3020         False       False         True      False                 NaN   
3021         False       False         True      False                 NaN   
3022         False        True        False      False                 NaN   

                                        Lat/Long  
0     POINT (-73.9561344937861 40.7940823884086)  
1     POINT (-73.9688574691102 40.7837825208444)  
2     POINT (-73.97428114848522 40.775533619083)  
3     POINT (-73.9596413903948 40.7903128889029)  
4     POINT (-73.9702676472613 40.7762126854894)  
...                                          ...  
3018  POINT (-73.9639431360458 40.7908677445466)  
3019  POINT (-73.9704015859639 40.7825600069973)  
3020  POINT (-73.9665871993517 40.7836775064883)  
3021  POINT (-73.9639941227864 40.7899152327912)  
3022  POINT (-73.9754794191553 40.7696404489025)  

[3023 rows x 31 columns]

Question 7:

1. You might prefer df.dropna() over del df['col'] if there are multiple rows or columns that have missing entries, as it can remove all rows or all columns with missing entries at once.

2. Using del df['col'] might be preferred over df.dropna() if there is a particular column that has an abundance of missing entries, making to faster to simply remove that troublesome column rather than have df.dropna() comb through the entire DataFrame

3. Applying del df['col'] before df.dropna() when using both could be important because removing troublesome columns that have several null entries first can reduce the amount of rows that will be removed due to df.dropna(). 

In [61]:
df.isna()

X      Y  Unique Squirrel ID  Hectare  Shift   Date  \
0     False  False               False    False  False  False   
1     False  False               False    False  False  False   
2     False  False               False    False  False  False   
3     False  False               False    False  False  False   
4     False  False               False    False  False  False   
...     ...    ...                 ...      ...    ...    ...   
3018  False  False               False    False  False  False   
3019  False  False               False    False  False  False   
3020  False  False               False    False  False  False   
3021  False  False               False    False  False  False   
3022  False  False               False    False  False  False   

      Hectare Squirrel Number    Age  Primary Fur Color  Highlight Fur Color  \
0                       False   True               True                 True   
1                       False   True               True                 True   
2                       False   True              False                 True   
3                       False  False              False                 True   
4                       False  False              False                False   
...                       ...    ...                ...                  ...   
3018                    False  False              False                 True   
3019                    False  False              False                False   
3020                    False  False              False                False   
3021                    False   True              False                False   
3022                    False  False              False                False   

      ...   Kuks  Quaas  Moans  Tail flags  Tail twitches  Approaches  \
0     ...  False  False  False       False          False       False   
1     ...  False  False  False       False          False       False   
2     ...  False  False  False       False          False       False   
3     ...  False  False  False       False          False       False   
4     ...  False  False  False       False          False       False   
...   ...    ...    ...    ...         ...            ...         ...   
3018  ...  False  False  False       False          False       False   
3019  ...  False  False  False       False          False       False   
3020  ...  False  False  False       False          False       False   
3021  ...  False  False  False       False          False       False   
3022  ...  False  False  False       False          False       False   

      Indifferent  Runs from  Other Interactions  Lat/Long  
0           False      False                True     False  
1           False      False                True     False  
2           False      False                True     False  
3           False      False                True     False  
4           False      False                True     False  
...           ...        ...                 ...       ...  
3018        False      False                True     False  
3019        False      False                True     False  
3020        False      False                True     False  
3021        False      False                True     False  
3022        False      False                True     False  

[3023 rows x 31 columns]

In [90]:
del df['Age']
del df['Primary Fur Color']
del df['Highlight Fur Color']
del df['Other Interactions']

In [91]:
df.isna()

X      Y  Unique Squirrel ID  Hectare  Shift   Date  \
0     False  False               False    False  False  False   
1     False  False               False    False  False  False   
2     False  False               False    False  False  False   
3     False  False               False    False  False  False   
4     False  False               False    False  False  False   
...     ...    ...                 ...      ...    ...    ...   
3018  False  False               False    False  False  False   
3019  False  False               False    False  False  False   
3020  False  False               False    False  False  False   
3021  False  False               False    False  False  False   
3022  False  False               False    False  False  False   

      Hectare Squirrel Number  Combination of Primary and Highlight Color  \
0                       False                                       False   
1                       False                                       False   
2                       False                                       False   
3                       False                                       False   
4                       False                                       False   
...                       ...                                         ...   
3018                    False                                       False   
3019                    False                                       False   
3020                    False                                       False   
3021                    False                                       False   
3022                    False                                       False   

      Color notes  Location  ...  Other Activities   Kuks  Quaas  Moans  \
0            True      True  ...              True  False  False  False   
1            True      True  ...              True  False  False  False   
2            True     False  ...              True  False  False  False   
3           False      True  ...              True  False  False  False   
4            True     False  ...              True  False  False  False   
...           ...       ...  ...               ...    ...    ...    ...   
3018         True     False  ...              True  False  False  False   
3019         True     False  ...              True  False  False  False   
3020         True     False  ...              True  False  False  False   
3021         True     False  ...              True  False  False  False   
3022         True     False  ...              True  False  False  False   

      Tail flags  Tail twitches  Approaches  Indifferent  Runs from  Lat/Long  
0          False          False       False        False      False     False  
1          False          False       False        False      False     False  
2          False          False       False        False      False     False  
3          False          False       False        False      False     False  
4          False          False       False        False      False     False  
...          ...            ...         ...          ...        ...       ...  
3018       False          False       False        False      False     False  
3019       False          False       False        False      False     False  
3020       False          False       False        False      False     False  
3021       False          False       False        False      False     False  
3022       False          False       False        False      False     False  

[3023 rows x 27 columns]

In [92]:
del df['Color notes']
del df['Location']
del df['Other Activities']

In [93]:
df.isna()

X      Y  Unique Squirrel ID  Hectare  Shift   Date  \
0     False  False               False    False  False  False   
1     False  False               False    False  False  False   
2     False  False               False    False  False  False   
3     False  False               False    False  False  False   
4     False  False               False    False  False  False   
...     ...    ...                 ...      ...    ...    ...   
3018  False  False               False    False  False  False   
3019  False  False               False    False  False  False   
3020  False  False               False    False  False  False   
3021  False  False               False    False  False  False   
3022  False  False               False    False  False  False   

      Hectare Squirrel Number  Combination of Primary and Highlight Color  \
0                       False                                       False   
1                       False                                       False   
2                       False                                       False   
3                       False                                       False   
4                       False                                       False   
...                       ...                                         ...   
3018                    False                                       False   
3019                    False                                       False   
3020                    False                                       False   
3021                    False                                       False   
3022                    False                                       False   

      Above Ground Sighter Measurement  Specific Location  ...  Foraging  \
0                                 True               True  ...     False   
1                                 True               True  ...     False   
2                                False               True  ...     False   
3                                 True               True  ...     False   
4                                 True              False  ...     False   
...                                ...                ...  ...       ...   
3018                             False               True  ...     False   
3019                             False               True  ...     False   
3020                             False               True  ...     False   
3021                             False               True  ...     False   
3022                             False               True  ...     False   

       Kuks  Quaas  Moans  Tail flags  Tail twitches  Approaches  Indifferent  \
0     False  False  False       False          False       False        False   
1     False  False  False       False          False       False        False   
2     False  False  False       False          False       False        False   
3     False  False  False       False          False       False        False   
4     False  False  False       False          False       False        False   
...     ...    ...    ...         ...            ...         ...          ...   
3018  False  False  False       False          False       False        False   
3019  False  False  False       False          False       False        False   
3020  False  False  False       False          False       False        False   
3021  False  False  False       False          False       False        False   
3022  False  False  False       False          False       False        False   

      Runs from  Lat/Long  
0         False     False  
1         False     False  
2         False     False  
3         False     False  
4         False     False  
...         ...       ...  
3018      False     False  
3019      False     False  
3020      False     False  
3021      False     False  
3022      False     False  

[3023 rows x 24 columns]

In [94]:
del df['Above Ground Sighter Measurement']
del df['Specific Location']

In [95]:
df.isna()

X      Y  Unique Squirrel ID  Hectare  Shift   Date  \
0     False  False               False    False  False  False   
1     False  False               False    False  False  False   
2     False  False               False    False  False  False   
3     False  False               False    False  False  False   
4     False  False               False    False  False  False   
...     ...    ...                 ...      ...    ...    ...   
3018  False  False               False    False  False  False   
3019  False  False               False    False  False  False   
3020  False  False               False    False  False  False   
3021  False  False               False    False  False  False   
3022  False  False               False    False  False  False   

      Hectare Squirrel Number  Combination of Primary and Highlight Color  \
0                       False                                       False   
1                       False                                       False   
2                       False                                       False   
3                       False                                       False   
4                       False                                       False   
...                       ...                                         ...   
3018                    False                                       False   
3019                    False                                       False   
3020                    False                                       False   
3021                    False                                       False   
3022                    False                                       False   

      Running  Chasing  ...  Foraging   Kuks  Quaas  Moans  Tail flags  \
0       False    False  ...     False  False  False  False       False   
1       False    False  ...     False  False  False  False       False   
2       False    False  ...     False  False  False  False       False   
3       False    False  ...     False  False  False  False       False   
4       False    False  ...     False  False  False  False       False   
...       ...      ...  ...       ...    ...    ...    ...         ...   
3018    False    False  ...     False  False  False  False       False   
3019    False    False  ...     False  False  False  False       False   
3020    False    False  ...     False  False  False  False       False   
3021    False    False  ...     False  False  False  False       False   
3022    False    False  ...     False  False  False  False       False   

      Tail twitches  Approaches  Indifferent  Runs from  Lat/Long  
0             False       False        False      False     False  
1             False       False        False      False     False  
2             False       False        False      False     False  
3             False       False        False      False     False  
4             False       False        False      False     False  
...             ...         ...          ...        ...       ...  
3018          False       False        False      False     False  
3019          False       False        False      False     False  
3020          False       False        False      False     False  
3021          False       False        False      False     False  
3022          False       False        False      False     False  

[3023 rows x 22 columns]

In [96]:
df = df.dropna()

In [97]:
df

X          Y Unique Squirrel ID Hectare Shift      Date  \
0    -73.956134  40.794082     37F-PM-1014-03     37F    PM  10142018   
1    -73.968857  40.783783     21B-AM-1019-04     21B    AM  10192018   
2    -73.974281  40.775534     11B-PM-1014-08     11B    PM  10142018   
3    -73.959641  40.790313     32E-PM-1017-14     32E    PM  10172018   
4    -73.970268  40.776213     13E-AM-1017-05     13E    AM  10172018   
...         ...        ...                ...     ...   ...       ...   
3018 -73.963943  40.790868     30B-AM-1007-04     30B    AM  10072018   
3019 -73.970402  40.782560     19A-PM-1013-05     19A    PM  10132018   
3020 -73.966587  40.783678     22D-PM-1012-07     22D    PM  10122018   
3021 -73.963994  40.789915     29B-PM-1010-02     29B    PM  10102018   
3022 -73.975479  40.769640      5E-PM-1012-01     05E    PM  10122018   

      Hectare Squirrel Number Combination of Primary and Highlight Color  \
0                           3                                          +   
1                           4                                          +   
2                           8                                      Gray+   
3                          14                                      Gray+   
4                           5                              Gray+Cinnamon   
...                       ...                                        ...   
3018                        4                                      Gray+   
3019                        5                                 Gray+White   
3020                        7                Gray+Black, Cinnamon, White   
3021                        2                       Gray+Cinnamon, White   
3022                        1                       Cinnamon+Gray, White   

      Running  Chasing  ...  Foraging   Kuks  Quaas  Moans  Tail flags  \
0       False    False  ...     False  False  False  False       False   
1       False    False  ...     False  False  False  False       False   
2       False     True  ...     False  False  False  False       False   
3       False    False  ...      True  False  False  False       False   
4       False    False  ...      True  False  False  False       False   
...       ...      ...  ...       ...    ...    ...    ...         ...   
3018    False    False  ...      True  False  False  False       False   
3019    False    False  ...      True  False  False  False       False   
3020    False    False  ...      True  False  False  False       False   
3021    False    False  ...     False  False  False  False       False   
3022    False    False  ...      True  False  False  False       False   

      Tail twitches  Approaches  Indifferent  Runs from  \
0             False       False        False      False   
1             False       False        False      False   
2             False       False        False      False   
3             False       False        False       True   
4             False       False        False      False   
...             ...         ...          ...        ...   
3018          False       False        False       True   
3019          False       False         True      False   
3020          False       False         True      False   
3021          False       False         True      False   
3022          False        True        False      False   

                                        Lat/Long  
0     POINT (-73.9561344937861 40.7940823884086)  
1     POINT (-73.9688574691102 40.7837825208444)  
2     POINT (-73.97428114848522 40.775533619083)  
3     POINT (-73.9596413903948 40.7903128889029)  
4     POINT (-73.9702676472613 40.7762126854894)  
...                                          ...  
3018  POINT (-73.9639431360458 40.7908677445466)  
3019  POINT (-73.9704015859639 40.7825600069973)  
3020  POINT (-73.9665871993517 40.7836775064883)  
3021  POINT (-73.9639941227864 40.7899152327912)  
3022  POINT (-73.9754794191553 40.7696404489025)  

[3023 rows x 

Question 7.4:

To remove all the missing data, I first took a look at the columns with visible missing entries using df.isna(), then I proceeded to move along and remove each visible troublesome column with del df['col']. Once I could no longer see null entries within the columns, I then finished by applying df.dropna() to remove any rows I could not see that had null entries.



Before:

In [98]:
df_original

X          Y Unique Squirrel ID Hectare Shift      Date  \
0    -73.956134  40.794082     37F-PM-1014-03     37F    PM  10142018   
1    -73.968857  40.783783     21B-AM-1019-04     21B    AM  10192018   
2    -73.974281  40.775534     11B-PM-1014-08     11B    PM  10142018   
3    -73.959641  40.790313     32E-PM-1017-14     32E    PM  10172018   
4    -73.970268  40.776213     13E-AM-1017-05     13E    AM  10172018   
...         ...        ...                ...     ...   ...       ...   
3018 -73.963943  40.790868     30B-AM-1007-04     30B    AM  10072018   
3019 -73.970402  40.782560     19A-PM-1013-05     19A    PM  10132018   
3020 -73.966587  40.783678     22D-PM-1012-07     22D    PM  10122018   
3021 -73.963994  40.789915     29B-PM-1010-02     29B    PM  10102018   
3022 -73.975479  40.769640      5E-PM-1012-01     05E    PM  10122018   

      Hectare Squirrel Number    Age Primary Fur Color  \
0                           3    NaN               NaN   
1                           4    NaN               NaN   
2                           8    NaN              Gray   
3                          14  Adult              Gray   
4                           5  Adult              Gray   
...                       ...    ...               ...   
3018                        4  Adult              Gray   
3019                        5  Adult              Gray   
3020                        7  Adult              Gray   
3021                        2    NaN              Gray   
3022                        1  Adult          Cinnamon   

         Highlight Fur Color  ...   Kuks  Quaas  Moans Tail flags  \
0                        NaN  ...  False  False  False      False   
1                        NaN  ...  False  False  False      False   
2                        NaN  ...  False  False  False      False   
3                        NaN  ...  False  False  False      False   
4                   Cinnamon  ...  False  False  False      False   
...                      ...  ...    ...    ...    ...        ...   
3018                     NaN  ...  False  False  False      False   
3019                   White  ...  False  False  False      False   
3020  Black, Cinnamon, White  ...  False  False  False      False   
3021         Cinnamon, White  ...  False  False  False      False   
3022             Gray, White  ...  False  False  False      False   

     Tail twitches  Approaches  Indifferent  Runs from  Other Interactions  \
0            False       False        False      False                 NaN   
1            False       False        False      False                 NaN   
2            False       False        False      False                 NaN   
3            False       False        False       True                 NaN   
4            False       False        False      False                 NaN   
...            ...         ...          ...        ...                 ...   
3018         False       False        False       True                 NaN   
3019         False       False         True      False                 NaN   
3020         False       False         True      False                 NaN   
3021         False       False         True      False                 NaN   
3022         False        True        False      False                 NaN   

                                        Lat/Long  
0     POINT (-73.9561344937861 40.7940823884086)  
1     POINT (-73.9688574691102 40.7837825208444)  
2     POINT (-73.97428114848522 40.775533619083)  
3     POINT (-73.9596413903948 40.7903128889029)  
4     POINT (-73.9702676472613 40.7762126854894)  
...                                          ...  
3018  POINT (-73.9639431360458 40.7908677445466)  
3019  POINT (-73.9704015859639 40.7825600069973)  
3020  POINT (-73.9665871993517 40.7836775064883)  
3021  POINT (-73.9639941227864 40.7899152327912)  
3022  POINT (-73.9754794191553 40.7696404489025)  

[3023 rows x 31 columns]

After

In [99]:
df

X          Y Unique Squirrel ID Hectare Shift      Date  \
0    -73.956134  40.794082     37F-PM-1014-03     37F    PM  10142018   
1    -73.968857  40.783783     21B-AM-1019-04     21B    AM  10192018   
2    -73.974281  40.775534     11B-PM-1014-08     11B    PM  10142018   
3    -73.959641  40.790313     32E-PM-1017-14     32E    PM  10172018   
4    -73.970268  40.776213     13E-AM-1017-05     13E    AM  10172018   
...         ...        ...                ...     ...   ...       ...   
3018 -73.963943  40.790868     30B-AM-1007-04     30B    AM  10072018   
3019 -73.970402  40.782560     19A-PM-1013-05     19A    PM  10132018   
3020 -73.966587  40.783678     22D-PM-1012-07     22D    PM  10122018   
3021 -73.963994  40.789915     29B-PM-1010-02     29B    PM  10102018   
3022 -73.975479  40.769640      5E-PM-1012-01     05E    PM  10122018   

      Hectare Squirrel Number Combination of Primary and Highlight Color  \
0                           3                                          +   
1                           4                                          +   
2                           8                                      Gray+   
3                          14                                      Gray+   
4                           5                              Gray+Cinnamon   
...                       ...                                        ...   
3018                        4                                      Gray+   
3019                        5                                 Gray+White   
3020                        7                Gray+Black, Cinnamon, White   
3021                        2                       Gray+Cinnamon, White   
3022                        1                       Cinnamon+Gray, White   

      Running  Chasing  ...  Foraging   Kuks  Quaas  Moans  Tail flags  \
0       False    False  ...     False  False  False  False       False   
1       False    False  ...     False  False  False  False       False   
2       False     True  ...     False  False  False  False       False   
3       False    False  ...      True  False  False  False       False   
4       False    False  ...      True  False  False  False       False   
...       ...      ...  ...       ...    ...    ...    ...         ...   
3018    False    False  ...      True  False  False  False       False   
3019    False    False  ...      True  False  False  False       False   
3020    False    False  ...      True  False  False  False       False   
3021    False    False  ...     False  False  False  False       False   
3022    False    False  ...      True  False  False  False       False   

      Tail twitches  Approaches  Indifferent  Runs from  \
0             False       False        False      False   
1             False       False        False      False   
2             False       False        False      False   
3             False       False        False       True   
4             False       False        False      False   
...             ...         ...          ...        ...   
3018          False       False        False       True   
3019          False       False         True      False   
3020          False       False         True      False   
3021          False       False         True      False   
3022          False        True        False      False   

                                        Lat/Long  
0     POINT (-73.9561344937861 40.7940823884086)  
1     POINT (-73.9688574691102 40.7837825208444)  
2     POINT (-73.97428114848522 40.775533619083)  
3     POINT (-73.9596413903948 40.7903128889029)  
4     POINT (-73.9702676472613 40.7762126854894)  
...                                          ...  
3018  POINT (-73.9639431360458 40.7908677445466)  
3019  POINT (-73.9704015859639 40.7825600069973)  
3020  POINT (-73.9665871993517 40.7836775064883)  
3021  POINT (-73.9639941227864 40.7899152327912)  
3022  POINT (-73.9754794191553 40.7696404489025)  

[3023 rows x 

ChatBot Session 2: 
    
Here's a summary of our chat:

describe() Method: We discussed the describe() method in Pandas, which provides descriptive statistics for each column in a DataFrame. The output includes count, mean, standard deviation, min, 25th percentile, median (50th percentile), 75th percentile, and max values.

del df['col']: We covered how del df['col'] removes a specific column from a DataFrame. This operation is straightforward but less flexible compared to handling missing data.

Utilizing Missing Data: We explored how dropna() can be used to handle missing data by removing rows or columns with missing values. Additionally, dropna() can be combined with imputation techniques or used conditionally to clean your data while retaining useful information.

Removing Multiple Columns: We discussed how to remove multiple columns using del, which involves iterating over a list of column names. We also mentioned an alternative approach using df.drop(columns=columns_to_remove), which is more concise and efficient for removing multiple columns.

Let me know if there's anything you'd like to add or clarify!

https://chatgpt.com/share/a74a0cf8-c415-49b2-a217-47981c50d7ac

Question 8:

In [4]:
df = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv')
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [7]:
df.groupby('embark_town')['fare'].describe()

count       mean        std     min       25%    50%        75%  \
embark_town                                                                    
Cherbourg    168.0  59.954144  83.912994  4.0125  13.69795  29.70  78.500025   
Queenstown    77.0  13.276030  14.188047  6.7500   7.75000   7.75  15.500000   
Southampton  644.0  27.079812  35.887993  0.0000   8.05000  13.00  27.900000   

                  max  
embark_town            
Cherbourg    512.3292  
Queenstown    90.0000  
Southampton  263.0000

Question 8.1:

df.groupby("col1")["col2"].describe() organizes the describe() method to have the rows from the given col1 to act as row headers for the values in the given col2. For example, if you wanted to see what the average fare of the Titanic from a particular town was or how expensive the fare in a certain town was was, you can setup to have the town embarked be the row categories, while the fares are the numerical values being processed.

In [6]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

Question 8.2:

The reason why the values of count in something like df.groupby("col1")["col2"].describe() differ so much from the regular df.describe() is because of what is being counted. df.describe() counts all instances of non-null values for each column. df.groupby("col1")["col2"].describe() on the other hand counts the amount of non-null entries in column 2 in relation to the non-null attributes of column 1. While, for example, there can be 891 non-null entries in the fare column, there are only 889 non-null entries in the fare column in relation to the filled entries of the embark_town column.

In [1]:
df = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv')
df

NameError: name 'pd' is not defined

In [2]:
import pandas as pd

df = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv')
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

Question 8.3;A

ChatGPT took less effort

In [3]:
df = pd.read_csvdf = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv')
df

HTTPError: HTTP Error 404: Not Found

In [4]:
import pandas as pd  # Make sure pandas is imported

# Correct URL for the Titanic dataset
df = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv')

# Display the DataFrame to confirm it loaded correctly
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [5]:
df = pd.read_csvdf = pd.read_csv('titanics.csv')
df

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

In [6]:
import pandas as pd  # Import pandas library

# Load the Titanic dataset from the URL
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

# Display the first few rows to confirm the data loaded correctly
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

Question 8.3;B

ChatGPT still more efficient, multiple causes for HTTP and FileNotFound errors on google, less straight-forward

In [7]:
df.groupby('sex')['fare'].describe()

count       mean        std   min        25%   50%    75%       max
sex                                                                        
female  314.0  44.479818  57.997698  6.75  12.071875  23.0  55.00  512.3292
male    577.0  25.523893  43.138263  0.00   7.895800  10.5  26.55  512.3292

In [8]:
DF.groupby('sex')['fare'].describe()

NameError: name 'DF' is not defined

In [9]:
# Group by 'sex' and describe the 'fare' column
df.groupby('sex')['fare'].describe()

count       mean        std   min        25%   50%    75%       max
sex                                                                        
female  314.0  44.479818  57.997698  6.75  12.071875  23.0  55.00  512.3292
male    577.0  25.523893  43.138263  0.00   7.895800  10.5  26.55  512.3292

8.3;C

ChatGPT wins again, StackOverflow does not share my plight.

In [10]:
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csvurl)

df.head()

SyntaxError: unmatched ')' (1858165749.py, line 2)

In [11]:
# Load the Titanic dataset from the URL
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)  # Corrected this line

# Display the first few rows to confirm the data loaded correctly
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

8.3;D

ChatGPT better, but I believe that even the tiniest bit of familiarity of Python could have easily corrected that mistake faster than it takes to copy and paste into ChatGPT

In [12]:
df.groupby('sex')['fare'].describle()

AttributeError: 'SeriesGroupBy' object has no attribute 'describle'

In [13]:
# Group by 'sex' and describe the 'fare' column
df.groupby('sex')['fare'].describe()

count       mean        std   min        25%   50%    75%       max
sex                                                                        
female  314.0  44.479818  57.997698  6.75  12.071875  23.0  55.00  512.3292
male    577.0  25.523893  43.138263  0.00   7.895800  10.5  26.55  512.3292

8.3;E

Google does not provide enough help, ChatGPT unfortunately (or fortunately) was more responsive, as in it took less than 5 seconds.

In [14]:
df.groupby('Sex')['fare'].describe()

KeyError: 'Sex'

In [15]:
# Group by 'sex' (lowercase) and describe the 'fare' column
df.groupby('sex')['fare'].describe()

count       mean        std   min        25%   50%    75%       max
sex                                                                        
female  314.0  44.479818  57.997698  6.75  12.071875  23.0  55.00  512.3292
male    577.0  25.523893  43.138263  0.00   7.895800  10.5  26.55  512.3292

8.3;F

Google provides adequate information to troubleshoot, but ChatGPT is more straightforward, as in it outright tells you the solution

In [16]:
df.groupby(sex)['fare'].describe()

NameError: name 'sex' is not defined

In [17]:
# Group by 'sex' (string literal) and describe the 'fare' column
df.groupby('sex')['fare'].describe()

count       mean        std   min        25%   50%    75%       max
sex                                                                        
female  314.0  44.479818  57.997698  6.75  12.071875  23.0  55.00  512.3292
male    577.0  25.523893  43.138263  0.00   7.895800  10.5  26.55  512.3292

8.3;G

Google searches eventually lead to a solution, but ChatGPT provide you the reason why the NameError occurs, and how to fix it quickly (most likely has to do with chat history).

ChatBot Session 3:

Here's a summary of our session:

Titanic Dataset Analysis:

You wanted to perform basic summary analyses on the Titanic dataset using code involving groupby and describe.
We discussed the functionality of df.groupby("col1")["col2"].describe(), which groups the DataFrame by "col1" and calculates summary statistics for "col2" within each group.
Code Troubleshooting:

Loading the Dataset: You initially encountered issues with loading the dataset due to typos in the URL and file name. Correct code was provided for loading the dataset from the URL: df = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv').
Using the DataFrame: You encountered errors related to variable names (DF vs. df) and typos in method names (describle instead of describe).
Column Names: The errors also included issues with case sensitivity of column names ('sex' vs. 'Sex').
Correct Code for Analysis:

The correct code for grouping and describing the 'fare' column by 'sex' in the Titanic dataset is:
python
Copy code
import pandas as pd
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)
df.groupby('sex')['fare'].describe()
If you have any more questions or need further clarification, feel free to ask!

https://chatgpt.com/share/b799d679-8164-4744-a934-b3e0ec7898ff